### PR TITLE
Run the linters etc from overcommit

### PR DESCRIPTION
**Pull request by Br3nda**
This will run the linters from overcommit on travis. Devs can also choose to run overcommit on their git checkouts, which will warn/fail if you try to commit/push code that's just gonna fail in the travis build anyways.

I've enabled our existing linters as fail_on_warn
and i've set the other as warn_on_fail

Note: i had to cleanup the hardtabs in one file

### CODE REVIEW

**Comment by Br3nda** [comment]
I'll push some changes rubocop doesn't like, to check it will fail.

**Comment by Br3nda** [comment]
this failed with "Looks like you committed changes to Gemfile but not Gemfile.lock"
which is true - i sorted Gemfile to please rubocop

**Comment by Br3nda** [comment]
Linters running happily. Some with warnings because we're not compliant yet

```
Running pre-commit hooks
Check if database schema is up to date..........[RailsSchemaUpToDate] OK
Analyze with ESLint..........................................[EsLint] OK
Check for trailing whitespace....................[TrailingWhitespace] OK
Check for hard tabs........................................[HardTabs] OK
Check for broken symlinks............................[BrokenSymlinks] OK
Check for case-insensitivity conflicts................[CaseConflicts] OK
Check for merge conflicts............................[MergeConflicts] OK
Analyze with csslint........................................[CssLint] WARNING
Warnings on modified lines:
/home/travis/build/Growstuff/growstuff/app/assets/stylesheets/bootstrap-accessibility.css: line 1, col 215, Warning - Rule doesn't have all its properties in alphabetical order. (order-alphabetical)
/home/travis/build/Growstuff/growstuff/app/assets/stylesheets/bootstrap-accessibility.css: line 1, col 681, Warning - Rule doesn't have all its properties in alphabetical order. (order-alphabetical)
/home/travis/build/Growstuff/growstuff/spec/javascripts/support/jasmine.css: line 5, col 1, Warning - Don't use IDs in selectors. (ids)
/home/travis/build/Growstuff/growstuff/spec/javascripts/support/jasmine.css: line 5, col 1, Warning - Rule doesn't have all its properties in alphabetical order. (order-alphabetical)
Analyze with coffeelint..................................[CoffeeLint] WARNING
Errors on modified lines:
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/plantings.js.coffee:3:error Line exceeds maximum allowed length Length is 83, max is 80.
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/seeds.js.coffee:3:error Line exceeds maximum allowed length Length is 83, max is 80.
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/seeds.js.coffee:13:error Line exceeds maximum allowed length Length is 342, max is 80.
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/seeds.js.coffee:29:error Line exceeds maximum allowed length Length is 340, max is 80.
Analyze with JSHint..........................................[JsHint] WARNING
Warnings on modified lines:
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/graphs/bar_label_group.js: line 34, col 33, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/graphs/bar_label_group.js: line 40, col 5, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/app/assets/javascripts/graphs/width_scale.js: line 29, col 4, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/spec/javascripts/graphs/bar_group_spec.js: line 34, col 35, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/spec/javascripts/graphs/bar_label_group_spec.js: line 23, col 7, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/spec/javascripts/graphs/bar_label_group_spec.js: line 26, col 41, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/spec/javascripts/graphs/bar_label_group_spec.js: line 31, col 7, Missing semicolon. (W033)
/home/travis/build/Growstuff/growstuff/spec/javascripts/graphs/bar_label_group_spec.js: line 38, col 7, Missing semicolon. (W033)
List installed gems with newer versions available....[BundleOutdated] WARNING
Warnings:
You are trying to check outdated gems in deployment mode. Run `bundle outdated`
elsewhere.
If this is a development machine, remove the
/home/travis/build/Growstuff/growstuff/Gemfile freeze
by running `bundle install --no-deployment`.
Analyze with RuboCop........................................[RuboCop] OK
Analyze with haml-lint.....................................[HamlLint] OK```

File: .travis.yml [hunk 1]
```diff
@@ -7,49 +7,55 @@ cache:
...
-      bundle exec rake db:create db:migrate db:test:prepare
-      bundle exec rake assets:precompile
-    fi
+    set -e;
+    if [ "${STATIC_CHECKS}" = "true" ]; then
+      gem install overcommit rubocop haml-lint;
+      overcommit --install;
+      overcommit --sign;
+      overcommit --sign pre-commit;
+      # pip install yamllint;
```
Comment (pozorvlak) [inline .travis.yml]: I'd rather we don't commit commented-out code: I guess the logic here is that if we enable yamllint we should add this line here?

File: .travis.yml [hunk 2]
```diff
@@ -7,49 +7,55 @@ cache:
...
+      npm install -g eslint jshint coffeelint csslint;
```
Comment (pozorvlak) [inline .travis.yml]: Our dependencies are now split across `Gemfile` and `.travis.yml`. Are there gems we can depend on wrapping these tools?

File: script/check_static [hunk 3]
```diff
@@ -3,9 +3,7 @@
 checks = [
   'script/check_gemfile',
   'script/check_contributors_md',
-  'script/check_conflicts',
-  'rubocop --display-cop-names --rails',
-  "haml-lint app/views"
+  'script/check_conflicts'
```
Comment (pozorvlak) [inline script/check_static]: This defeats the idea of "you can run all the static checks with one command". Please either move the call to `overcommit -r` in here, or kill this script and add the ad-hoc checks to overcommit (and change the "how to run all the tests" instructions in `CONTRIBUTING.md` accordingly).

File: script/install_phantomjs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (pozorvlak) [inline script/install_phantomjs]: Good call on moving this code into a script.

File: .travis.yml (hunk 1 above)
Comment (Br3nda) [inline .travis.yml]: I think our options are adding virtualenv to hold this python requirement, or moving our travis to allow sudo pip install.

Comment (Br3nda) [inline .travis.yml]: or rather, i'm really keen on tunning yaml lint (there were a few bogus yaml files i cleaned up months ago) --- but can't until we're able install python modules within travis

**Comment by pozorvlak** [comment]
It looks like we can pass the --user flag to `pip install`:
https://docs.travis-ci.com/user/languages/python/#pip

**Comment by pozorvlak** [comment]
I now see that overcommit runs linters via Git commit hooks (hence the
name) so users won't need to run `overcommit -r` explicitly. Sorry about
that. OTOH, please document `overcommit --install && overcommit --sign` in
CONTRIBUTING.md. We should maybe move our project-specific static checks
into overcommit hooks, but I guess that can be done later.

**Comment by pozorvlak** [comment]
There's also a yaml-lint ruby gem, which looks like it could be invoked as
a script: https://rubygems.org/gems/yaml-lint/versions/0.0.7 Less capable
than the Python one, AFAICT, but it would be nice to avoid having to deal
with Yet Another Packaging System.

FWIW, it's possible to install yamllint (the Python one) via apt-get, but
it's not included in the whitelist of packages that are allowed in Travis's
container infrastructure.

**Comment by Br3nda** [comment]
This is failing cos:
Check for trailing whitespace....................[TrailingWhitespace] FAILED
Errors on modified lines:
/home/travis/build/Growstuff/growstuff/Gemfile:66:gem 'omniauth', '~> 1.3' 

Will look at that soon. also pip --user sounds good.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -24,6 +23,12 @@ checks before being merged. You can run tests locally as follows:
    failing tests so you can easily re-run particular ones.
  - `rspec --only-failures` to re-run all Ruby tests that failed last time.
 
+ Growstuff runs several linters and checkers before a change is merge. These run
```
Comment (pozorvlak) [inline CONTRIBUTING.md]: Typo: "merged".

**Pull request merged**